### ISSUE TITLE
can we extend comparisons to none to allow other types?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#17255]
Consider this example, which I wrote without really thinking about it when updating something else:

``` chapel
var x: int;

if x == none {
  writeln("it is none");
} else {
  writeln("it is not none");
}
```

It results in a type error because we can't compare `none` and `int`.

However since `none` is the only value of the type `nothing`, it seems to me that any value of any type other than `nothing` (note 1) should compare as not equal.

note 1: any type other than `nothing` or potentially `dtAny` if it is possible to get at that in user code.

**Comment by bradcray** [comment]
I'm not categorically opposed to this, but trying to understand where we'd draw the line.  For example, we could define a catch-all cross-type overload of `==` like:

```chapel
pragma "last resort"
proc ==(x: ?t1, y: ?t2) where t1 != t2 {
  return false;
}
```

but don't (currently at least).  Is there something that makes supporting cross-type comparisons with `none` more compelling than for other types?

**Comment by mppf** [comment]
>  Is there something that makes supporting cross-type comparisons with none more compelling than for other types?

Yeah. It only has one value - so it's not particularly useful to talk about `==` on it unless it can compare across types.

I don't think we have to do this or anything, it's just something I assumed would work and then did not. Sometimes that leads to an interesting language design improvement. I didn't see or recall discussion of it before.

It came up when I was using a `none` value as a placeholder for a lack of a value in a generic function, like this:

``` chapel
proc doSomethingWithTwoThings( thing1, thing2 ) {
  if thing1 != none {
     for item in thing1 { ... }
  }
  if thing2 != none {
    for item in thing2 { ... }
  }
}
doSomethingWithTwoThings( [1,2], [3,4] );

doSomethingWithTwoThings( [5,6], none );
```

You could view that as a context where there is a kind of optional argument in a generic, which is something `none`/`nothing` is intended to help with. I can write `if !isNothingType(thing1.type)` but that is a bit of a mouthful.
